File: pkg/profiling/continuous/checker/bpf/network/network.go
// Licensed to Apache Software Foundation (ASF) under one or more contributor
// license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright
// ownership. Apache Software Foundation (ASF) licenses this file to you under
// the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package network

import (
	"os"
	"sync"
	"time"

	"github.com/hashicorp/go-multierror"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"

	"github.com/apache/skywalking-rover/pkg/logger"
	"github.com/apache/skywalking-rover/pkg/tools/btf"
)

// $BPF_CLANG and $BPF_CFLAGS are set by the Makefile.
// nolint
//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -no-global-types -target bpfel -cc $BPF_CLANG -cflags $BPF_CFLAGS bpf $REPO_ROOT/bpf/profiling/continuous/network.c -- -I$REPO_ROOT/bpf/include -D__TARGET_ARCH_x86

var log = logger.GetLogger("profiling", "continuous", "checker", "network", "bpf")

var locker sync.Mutex
var bpf *bpfObjects
var bpfLinker *btf.Linker
var monitoringProcesses map[int32]*monitoringProcessInfo
var notifiers []EventNotify

func init() {
	monitoringProcesses = make(map[int32]*monitoringProcessInfo)
}

type BufferEvent interface {
	Pid() int32
	RequestURI() string
	IsResponseError() bool
	Duration() time.Duration
	StartTime() time.Time
}

type EventNotify interface {
	ReceiveBufferEvent(event BufferEvent)
}

func AddEventNotify(notify EventNotify) {
	notifiers = append(notifiers, notify)
}

// AddWatchProcess add the process to monitoring with from source
// BPF would be start automatically
func AddWatchProcess(pid int32, from string) error {
	if e := startBPFIfNeed(); e != nil {
		return e
	}

	// adding to the cache
	monitoring := monitoringProcesses[pid]
	if monitoring == nil {
		monitoring = newMonitoringProcessInfo(pid)
		monitoringProcesses[pid] = monitoring
	}

	// start monitoring process
	if e := monitoring.AddSource(from); e != nil {
		// remove the source if add failure
		if deleteProcess, _ := monitoring.RemoveSource(from); deleteProcess {
			delete(monitoringProcesses, pid)
		}
		return e
	}
	return nil
}

// RemoveWatchProcess remove the process to monitoring with from source
// if no process is monitoring, the BPF would be stopped
func RemoveWatchProcess(pid int32, from string) error {
	defer func() {
		if err := shutdownBPFIfNoProcesses(); err != nil {
			log.Warnf("shutdown the BPF failure: %v", err)
		}
	}()
	monitoring := monitoringProcesses[pid]
	if monitoring == nil {
		return nil
	}

	deleteProcess, err := monitoring.RemoveSource(from)
	if deleteProcess {
		delete(monitoringProcesses, pid)
	}
	return err
}

func ForceShutdownBPF() error {
	// shutdown all processes
	var err error
	for _, p := range monitoringProcesses {
		if e := p.Shutdown(); e != nil {
			err = multierror.Append(err, e)
		}
	}
	// shutdown the main BPF
	monitoringProcesses = make(map[int32]*monitoringProcessInfo)
	if e := shutdownBPF(); e != nil {
		err = multierror.Append(err, e)
	}
	return err
}

// start the BPF program if contains process that needs monitor
func startBPFIfNeed() error {
	locker.Lock()
	defer locker.Unlock()

	if bpf != nil {
		return nil
	}

	bpf = &bpfObjects{}
	if err := loadBpfObjects(bpf, btf.GetEBPFCollectionOptionsIfNeed()); err != nil {
		return err
	}
	bpfLinker = btf.NewLinker()
	bpfLinker.AddLink(link.Kprobe, map[string]*ebpf.Program{"tcp_sendmsg": bpf.TcpSendmsg})
	bpfLinker.AddLink(link.Kprobe, map[string]*ebpf.Program{"tcp_recvmsg": bpf.TcpRecvmsg})
	bpfLinker.AddLink(link.Kretprobe, map[string]*ebpf.Program{"tcp_recvmsg": bpf.RetTcpRecvmsg})

	reader := newNetworkBufferReader(func(event BufferEvent) {
		for _, n := range notifiers {
			n.ReceiveBufferEvent(event)
		}
	})
	bpfLinker.ReadEventAsyncWithBufferSize(bpf.SocketBufferSendQueue, reader.Read, os.Getpagesize()*100, reader.BufferDataBPFSupplier)

	if err := bpfLinker.HasError(); err != nil {
		_ = bpfLinker.Close()
		_ = bpf.Close()
		bpf = nil
		bpfLinker = nil
		return err
	}
	return nil
}

// shutdown the BPF program if we don't contain any process need to be monitored
func shutdownBPFIfNoProcesses() error {
	// if still contains monitoring process then the BPF should not be stopped
	if len(monitoringProcesses) != 0 {
		return nil
	}
	return shutdownBPF()
}

func shutdownBPF() error {
	locker.Lock()
	defer locker.Unlock()

	if bpf == nil {
		return nil
	}
	// if still contains monitoring process then the BPF should not be stopped
	if len(monitoringProcesses) != 0 {
		return nil
	}

	var err error
	if e := bpfLinker.Close(); e != nil {
		err = multierror.Append(err, e)
	}
	if e := bpf.Close(); e != nil {
		err = multierror.Append(err, e)
	}
	bpf = nil
	return err
}

File: bpf/profiling/continuous/network.c
// Licensed to Apache Software Foundation (ASF) under one or more contributor
// license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright
// ownership. Apache Software Foundation (ASF) licenses this file to you under
// the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <linux/sched.h>
#include <linux/bpf.h>
#include <linux/ptrace.h>
#include <asm/ptrace.h>
#include <asm/errno.h>
#include <asm/socket.h>
#include <linux/netfilter_ipv4.h>
#include <linux/tcp.h>
#include <bpf/bpf_endian.h>
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>
#include <bpf/bpf_core_read.h>
#include "network.h"
#include "protocol_analyzer.h"

char __license[] SEC("license") = "Dual MIT/GPL";

static __always_inline bool socket_should_trace(__u64 id, struct sock *sock) {
    // check the pid is monitoring
    __u32 tgid = (__u32)(id >> 32);
    if (tgid_should_trace(tgid) == false) {
        return false;
    }

    // check the socket if server side
    int max_ack_backlog = 0;
    if (0 != bpf_core_read(&max_ack_backlog, sizeof(max_ack_backlog),
                &sock->sk_max_ack_backlog)) {
        return true;
    }
    if (max_ack_backlog == 0) {
        return false;
    }
    return true;
}

static __always_inline void process_data(struct pt_regs *ctx, __u64 id, void *channel_ref, char *buf, __u64 size, __u64 timestamp) {
    if (size <= 0) {
        return;
    }

    if (size > MAX_PROTOCOL_SOCKET_READ_LENGTH) {
        size = MAX_PROTOCOL_SOCKET_READ_LENGTH;
    }
    __u32 kZero = 0;
    struct socket_buffer_reader_t* reader = bpf_map_lookup_elem(&socket_buffer_reader_map, &kZero);
    if (reader == NULL) {
        return;
    }
    asm volatile("%[size] &= 0xff;\n" ::[size] "+r"(size) :);
    bpf_probe_read(&reader->buffer, size & MAX_PROTOCOL_SOCKET_READ_LENGTH, buf);
    __u8 protocol;
    __u32 direction = analyze_protocol(reader->buffer, size & MAX_PROTOCOL_SOCKET_READ_LENGTH, &protocol);
    if (protocol != CONNECTION_PROTOCOL_UNKNOWN) {
        reader->timestamp = timestamp;
        reader->channel_ref = channel_ref;
        reader->pid = (__u32)(id >> 32);
        reader->protocol = protocol;
        reader->direction = (__u8)direction;
        reader->size = size & MAX_PROTOCOL_SOCKET_READ_LENGTH;
        bpf_perf_event_output(ctx, &socket_buffer_send_queue, BPF_F_CURRENT_CPU, reader, sizeof(*reader));
    }
}

static __always_inline void process_msghdr_data(struct pt_regs *ctx, __u64 id, void *channel_ref, struct msghdr *msg) {
    const struct iovec *iovec;
    iovec = _KERNEL(msg->msg_iter.iov);
    struct iovec iov;
    bpf_probe_read(&iov, sizeof(iov), iovec);
    char* buf = (char *)iov.iov_base;
    __u64 size = iov.iov_len;

    return process_data(ctx, id, channel_ref, buf, size, bpf_ktime_get_ns());
}

SEC("kprobe/tcp_sendmsg")
int tcp_sendmsg(struct pt_regs *ctx) {
    __u64 id = bpf_get_current_pid_tgid();
    struct sock *s = (void *)PT_REGS_PARM1(ctx);
    if (socket_should_trace(id, s) == false) {
        return 0;
    }

    struct msghdr *msg = (void *)PT_REGS_PARM2(ctx);
    process_msghdr_data(ctx, id, s, msg);
    return 0;
}

SEC("kprobe/tcp_recvmsg")
int tcp_recvmsg(struct pt_regs *ctx) {
    __u64 id = bpf_get_current_pid_tgid();
    struct sock *s = (void *)PT_REGS_PARM1(ctx);
    if (socket_should_trace(id, s) == false) {
        return 0;
    }

    struct msghdr *msg = (void *)PT_REGS_PARM2(ctx);
    struct recv_msg_args args = {};
    args.sock = s;
    args.msg = msg;
    bpf_map_update_elem(&receiving_args, &id, &args, 0);
    return 0;
}

SEC("kretprobe/tcp_recvmsg")
int ret_tcp_recvmsg(struct pt_regs *ctx) {
    __u64 id = bpf_get_current_pid_tgid();
    struct recv_msg_args *args = bpf_map_lookup_elem(&receiving_args, &id);
    int bytes_count = PT_REGS_RC(ctx);
    if (args != NULL && bytes_count > 0) {
        process_msghdr_data(ctx, id, args->sock, args->msg);
    }
    bpf_map_delete_elem(&receiving_args, &id);
    return 0;
}

#include "openssl.c"
#include "go_tls.c"
